// ui/component/GameControl.js
import React, { Component } from "react";
import { connect } from "react-redux";
import "./GameControl.css";
import Hammer from "react-hammerjs";

class GameControl extends Component {
  constructor(props) {
    super(props);
  }

  render() {
    const styles = {
      positions: {
        top: `${this.props.gamestate[this.props.gameControl]}%`
      }
    };
    const { positions } = styles;

    return (
      
        <div className="control-bar" style={positions}>
          <Hammer onPan={handleGameControl.bind(this)}>
            <div className="control-handle" />
          </Hammer>
          <div className="control-bar-bottom" />
          <div className="control-bar-middle" />
          <div className="control-bar-top" />
        </div>
    );
  }
}

function handleGameControl(event) {
  const deltaX = event.deltaX;
  const deltaY = event.deltaY;

  let newY = deltaY / 10;
  
  if (newY >= 0) {
    newY = 0;
  } else if (newY <= -20) {
    newY = -20;
  } 
  
  // this.setState({
  //   vertical: newY
  // });

  this.props.actions.updateGameControlY(this.props.gameControl, newY);
}

function mapStateToProps(state) {
  console.log('global', state);
  const { route, gamestate } = state;

  return {
    route, gamestate
  };
}

function mapDispatchToProps(dispatch) {
  return {
    actions: {
      changeRoute: routeName =>
        dispatch({
          type: "CHANGE_ROUTE",
          routeName
        }),
      updateGameControlY: (gameControl, newY) => 
        dispatch({
          type: "GAME_CONTROL_Y_UPDATE",
          gameControl: gameControl,
          newY: newY
        })  
    }
  };
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(GameControl);
