// ui/component/Home.js
import React, { Component } from "react";
import "./Home.css";
import { connect } from "react-redux";

import {
  BrowserRouter as Router,
  Redirect,
  Route,
  Link
} from "react-router-dom";

import NavCard from "ui/component/NavCard";
import Paper from "@material-ui/core/Paper";
import Grid from "@material-ui/core/Grid";

import ImageNewLobby from "ui/component/images/newlobby.jpg";
import ImageJoinGame from "ui/component/images/join.jpg";
import ImageLock from "ui/component/images/lock.png";

class Home extends Component {
  constructor(props) {
    super(props);
  }

  routeTo(target) {
    this.props.actions.changeRoute(target);
  }

  render() {
    if (this.props.route.name === "newlobby") {
      return <Redirect to="/newlobby" />;
    }
    if (this.props.route.name === "joingame") {
      return <Redirect to="/joingame" />;
    }

    return (
      <Grid container spacing={24} className="home-grid">
        <Grid item xs={8} className="home-grid-cell">
          <NavCard
            text="Join Game"
            image={ImageJoinGame}
            pos="nav-card-centerbottom"
            stretch="nav-card-cover"
            onClick={this.routeTo.bind(this, "joingame")}
          />
        </Grid>
        <Grid item xs={4} className="home-grid-cell">
          <NavCard
            text="Create Lobby"
            image={ImageNewLobby}
            pos="nav-card-centerbottom"
            stretch="nav-card-cover"
            onClick={this.routeTo.bind(this, "newlobby")}
          />
        </Grid>
        <Grid item xs={4} className="home-grid-cell">
          <NavCard
            text="Store"
            image={ImageLock}
            pos="nav-card-centercenter"
            stretch="nav-card-contain"
          />
        </Grid>
        <Grid item xs={4} className="home-grid-cell">
          <NavCard
            text="Equipments"
            image={ImageLock}
            pos="nav-card-centercenter"
            stretch="nav-card-contain"
          />
        </Grid>
        <Grid item xs={4} className="home-grid-cell">
          <NavCard
            text="Help"
            image={ImageLock}
            pos="nav-card-centercenter"
            stretch="nav-card-contain"
          />
        </Grid>
      </Grid>
    );
  }
}

function mapStateToProps(state) {
  const { player, route } = state;

  return {
    player,
    route
  };
}

function mapDispatchToProps(dispatch) {
  return {
    actions: {
      changeRoute: routeName =>
        dispatch({
          type: "CHANGE_ROUTE",
          routeName
        })
    }
  };
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Home);
