// ui/component/JoinGame.js
import React, { Component } from "react";
import "./JoinGame.css";
import { connect } from "react-redux";

import {
  BrowserRouter as Router,
  Redirect,
  Route,
  Link
} from "react-router-dom";

import Button from "@material-ui/core/Button";
import List from "@material-ui/core/List";
import ListItem from "@material-ui/core/ListItem";
import ListItemAvatar from "@material-ui/core/ListItemAvatar";
import ListItemSecondaryAction from "@material-ui/core/ListItemSecondaryAction";
import ListItemText from "@material-ui/core/ListItemText";
import Avatar from "@material-ui/core/Avatar";
import IconButton from "@material-ui/core/IconButton";
import FolderIcon from "@material-ui/icons/Folder";
import DeleteIcon from "@material-ui/icons/Delete";
import Divider from "@material-ui/core/Divider";

class JoinGame extends Component {
  constructor(props) {
    super(props);
  }

  componentDidMount() {
    this.props.actions.getRoomList();
  }

  routeTo(target) {
    this.props.actions.changeRoute(target);
  }

  render() {
    if (this.props.route.name === "home") {
      return <Redirect to="/" />;
    }
    if (this.props.route.name === "lobby") {
      return <Redirect to="/lobby" />;
    }

    return (
      <div className="joingame">
        <div className="joingame-list">
          <List>
            {this.props.rooms.map((room, index) => (
              <ListItem>
                <ListItemAvatar>
                  <Avatar
                    src={
                      "http://graph.facebook.com/" +
                      room.metadata.creator.fb_client_id +
                      "/picture"
                    }
                  />
                </ListItemAvatar>
                <ListItemText
                  primary={"Room: " + room.metadata.name + " Players: " + room.clients + "/4"}
                  secondary={room.metadata.creator.fb_client_name}
                />
                <ListItemSecondaryAction>
                  <IconButton aria-label="Delete">
                    <DeleteIcon />
                  </IconButton>
                </ListItemSecondaryAction>
              </ListItem>
            ))}
          </List>
        </div>

        <Divider />

        <div className="joingame-btngroup">
          <Button className="" onClick={this.routeTo.bind(this, "home")}>
            Back
          </Button>
        </div>
      </div>
    );
  }
}

function mapStateToProps(state) {
  const { player, rooms, route } = state;

  return {
    player,
    rooms,
    route
  };
}

function mapDispatchToProps(dispatch) {
  return {
    actions: {
      getRoomList: () =>
        dispatch({
          type: "GET_ROOM_LIST"
        }),
      changeRoute: routeName =>
        dispatch({
          type: "CHANGE_ROUTE",
          routeName
        })
    }
  };
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(JoinGame);
