// ui/component/JoinGame.js
import React, { Component } from "react";
import "./JoinGame.css";
import { connect } from "react-redux";

import {
  BrowserRouter as Router,
  Redirect,
  Route,
  Link
} from "react-router-dom";

import Button from "@material-ui/core/Button";

class JoinGame extends Component {
  constructor(props) {
    super(props);
  }

  componentDidMount() {
    this.props.actions.getRoomList();
  }

  routeTo(target) {
    this.props.actions.changeRoute(target);
  }

  render() {
    if (this.props.route.name === "home") {
      return <Redirect to="/" />;
    }
    if (this.props.route.name === "lobby") {
      return <Redirect to="/lobby" />;
    }

    return (
      <div className="joingame">
        <div className="joingame-list">
          {this.props.rooms.map((item, index) => (
            <div className="indent" key={index}>
              {item.id}
            </div>
          ))}
        </div>

        <div className="joingame-btngroup">
          <Button className="" onClick={this.routeTo.bind(this, "home")}>
            Back
          </Button>
        </div>
      </div>
    );
  }
}

function mapStateToProps(state) {
  const { rooms, route } = state;

  return {
    rooms,
    route
  };
}

function mapDispatchToProps(dispatch) {
  return {
    actions: {
      getRoomList: () =>
        dispatch({
          type: "GET_ROOM_LIST"
        }),
      changeRoute: routeName =>
        dispatch({
          type: "CHANGE_ROUTE",
          routeName
        })
    }
  };
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(JoinGame);
