// ui/component/Lobby.js
import React, { Component } from "react";
import "./Lobby.css";
import { connect } from "react-redux";
import Avatar from "@material-ui/core/Avatar";
import Badge from "@material-ui/core/Badge";
import TextField from "@material-ui/core/TextField";
import FormControlLabel from "@material-ui/core/FormControlLabel";
import Switch from "@material-ui/core/Switch";
import Button from "@material-ui/core/Button";
import Grid from "@material-ui/core/Grid";
import AccountCircle from "@material-ui/icons/AccountCircle";

class Lobby extends Component {
  constructor(props) {
    super(props);
  }

  checkPlayer(position) {
    console.log("??", position);
  }

  render() {
    return (
      <div className="lobby">
        <div className="lobby-main">
          <div className="lobby-main-left">
            <div className="lobby-main-left-position">
              <div
                className="lobby-pos lobby-pos-1"
                onClick={this.checkPlayer.bind(this, 1)}
              >
                <Badge badgeContent="&#9919;" color="primary">
                  <Avatar
                    className="lobby-avatar"
                    src={
                      "http://graph.facebook.com/" +
                      this.props.player.id +
                      "/picture"
                    }
                  />
                </Badge>
              </div>

              <div
                className="lobby-pos lobby-pos-2"
                onClick={this.checkPlayer.bind(this, 2)}
              >
                <Avatar
                  className="lobby-avatar"
                  src={
                    "http://graph.facebook.com/" +
                    this.props.player.id +
                    "/picture"
                  }
                />
              </div>

              <div
                className="lobby-pos lobby-pos-3"
                onClick={this.checkPlayer.bind(this, 3)}
              >
                <Avatar
                  className="lobby-avatar"
                  src={
                    "http://graph.facebook.com/" +
                    this.props.player.id +
                    "/picture"
                  }
                />
              </div>
              <div
                className="lobby-pos lobby-pos-4"
                onClick={this.checkPlayer.bind(this, 4)}
              >
                <Avatar
                  className="lobby-avatar"
                  src={
                    "http://graph.facebook.com/" +
                    this.props.player.id +
                    "/picture"
                  }
                />
              </div>
            </div>
          </div>

          <div className="lobby-main-right">
            <div>
              <Grid container alignItems="flex-end">
                <Grid item>
                  <AccountCircle />
                </Grid>
                <Grid item>
                  <TextField id="input-with-icon-grid" label="Chat" />
                </Grid>
              </Grid>
            </div>
          </div>
        </div>

        <div className="lobby-footer-action">
          <FormControlLabel
            control={<Switch classes="" disableRipple checked={true} />}
            label="Ready"
          />

          <Button className="">Exit</Button>
        </div>
      </div>
    );
  }
}

function mapStateToProps(state) {
  const { player } = state;

  return {
    player
  };
}

function mapDispatchToProps(dispatch) {
  return {
    actions: {
      joinRoom: (lobbyName, player) =>
        dispatch({
          type: "JOIN_ROOM",
          payload: {
            lobbyName,
            player
          }
        }),
      changeRoute: routeName =>
        dispatch({
          type: "CHANGE_ROUTE",
          routeName
        })
    }
  };
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Lobby);
