// ui/component/Main.js
import React, { Component } from "react";
import "./Main.css";
import { connect } from "react-redux";

import { BrowserRouter as Router, Route, Link } from "react-router-dom";
import MainAppBar from "ui/component/MainAppBar";
import Home from "ui/component/Home";
import NewLobby from "ui/component/NewLobby";
import JoinGame from "ui/component/JoinGame";
import Lobby from "ui/component/Lobby";
import Game from "ui/component/Game";

class Main extends Component {
  constructor(props) {
    super(props);

    const sPath = window.location.pathname;
    let sRoute = sPath.substring(1, sPath.length);
    sRoute = sRoute ? sRoute: "home";
    this.props.actions.changeRoute(sRoute);
  }

  render() {
    return (
      <div className="main">
        <MainAppBar />
        <Router>
          <div className="main-content">
            <Route exact path="/" component={Home} />
            <Route path="/joingame" component={JoinGame} />
            <Route path="/newlobby" component={NewLobby} />
            <Route path="/lobby" component={Lobby} />
            <Route path="/game" component={Game} />
          </div>
        </Router>
      </div>
    );
  }
}

function mapStateToProps(state) {
  const { route } = state;

  return {
    route
  };
}

function mapDispatchToProps(dispatch) {
  return {
    actions: {
      changeRoute: routeName =>
        dispatch({
          type: "CHANGE_ROUTE",
          routeName
        })
    }
  };
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Main);
