// ui/component/NewLobby.js
import React, { Component } from "react";
import "./NewLobby.css";
import { connect } from "react-redux";
import gameclient from "gameclient";
import { bindActionCreators } from "redux";

import { BrowserRouter as Router, Route, Link } from "react-router-dom";

import TextField from "@material-ui/core/TextField";
import Button from "@material-ui/core/Button";
class NewLobby extends Component {
  constructor(props) {
    super(props);

    this.state = {
      lobbyname: ""
    };
  }

  componentDidMount() {}

  createLobby() {
    const sLobbyName = this.state.lobbyname;
    this.props.actions.joinRoom(sLobbyName);
  }

  onCreateLobby(lobbyInfo) {
    console.log(lobbyInfo, "joined");
  }

  _handleTextFieldChange(e) {
    this.setState({
      lobbyname: e.target.value
    });
  }

  render() {
    return (
      <div className="newlobby">
        <div className="newlobby-form">
          <TextField
            label="Name"
            margin="normal"
            value={this.state.lobbyname}
            onChange={this._handleTextFieldChange.bind(this)}
          />
        </div>
        <div className="newlobby-btngroup">
          <Button
            variant="contained"
            color="primary"
            className=""
            onClick={this.createLobby.bind(this)}
          >
            Create
          </Button>
          <Button className="" component={Link} to="/">
            Cancel
          </Button>
        </div>
      </div>
    );
  }
}

function mapStateToProps(state) {
  const { player } = state;

  return {
    player
  };
}

function mapDispatchToProps(dispatch) {
  return {
    actions: {
      joinRoom: lobbyName =>
        dispatch({
          type: "JOIN_ROOM",
          payload: {
            lobbyName
          }
        })
    }
  };
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(NewLobby);
