// epics/game.js
import { Subject } from "rxjs";
import { mergeMap, filter } from "rxjs/operators";

import { Client } from "colyseus.js";
const client = new Client("ws://localhost:7000");
let room;

function handleRoomStateChange(subject, state) {
  switch (state.step) {
    case "NEW_ROOM":
      createNewRoom(subject, state);
      break;
    case "PLAYER_JOIN":
      joinRoom(subject, state);
      break;
    default:
      reportUnhandledStateChange(state);
  }
}

function createNewRoom(subject, state) {
  console.log("Room State NEW_ROOM", state);
  subject.next({
    type: "ROOM_STATE_INIT",
    payload: room.state
  });

  subject.next({
    type: "CHANGE_ROUTE",
    routeName: "lobby"
  });
}

function joinRoom(subject, state) {
  console.log("Room State PLAYER_JOIN", state);
}

function reportUnhandledStateChange(state) {
  console.log("Unhandled state change", state);
}

export const createRoomEpic = action$ =>
  action$.pipe(
    filter(action => action.type === "CREATE_ROOM"),

    mergeMap(action => {
      const subject = new Subject();
      room = client.join("standard", {
        standard: action.payload.lobbyName,
        scene: "bluesky",
        mode: "10Pts",
        client_id: "",
        client_session_id: "",
        fb_client_id: action.payload.player.id,
        fb_client_name: action.payload.player.name
      });

      room.onStateChange.add(state => {
        handleRoomStateChange(subject, state);
      });

      return subject;
    })
  );

export const joinRoomEpic = action$ =>
  action$.pipe(
    filter(action => action.type === "JOIN_ROOM"),

    mergeMap(action => {
      const subject = new Subject();
      room = client.join(action.payload.room.roomId, {
        standard: action.payload.room.metadata.name,
        client_id: "",
        client_session_id: "",
        fb_client_id: action.payload.player.id,
        fb_client_name: action.payload.player.name
      });

      room.onStateChange.add(state => {
        handleRoomStateChange(subject, state);
      });

      return subject;
    })
  );

export const getRoomListEpic = action$ =>
  action$.pipe(
    filter(action => action.type === "GET_ROOM_LIST"),

    mergeMap(action => {
      const subject = new Subject();
      // Call Colyseus client to get room list
      client.getAvailableRooms("standard", function(rooms, err) {
        if (err) {
          console.error(err);
        } else {
          console.log("Rooms from server", rooms);

          subject.next({
            type: "UPDATE_ROOM_LIST",
            rooms
          });
        }
      });
      return subject;
    })
  );
