// epics/game.js
import { Subject } from "rxjs";
import { mergeMap, filter } from "rxjs/operators";

import { Client } from "colyseus/colyseus.js";
const client = new Client("ws://localhost:7000");

export const createRoomEpic = action$ =>
  action$.pipe(
    filter(action => action.type === "CREATE_ROOM"),

    mergeMap(action => {
      const subject = new Subject();
      const room = client.join("standard", {
        standard: action.payload.lobbyName,
        scene: "bluesky",
        creator: {
          colyseus_client_id: client.id,
          fb_client_id: action.payload.player.id,
          fb_client_name: action.payload.player.name
        }
      });

      room.onJoin.add(() => {});

      room.onStateChange.addOnce(state => {
        console.log("Room State Init", state);
        subject.next({
          type: "ROOM_STATE_INIT",
          payload: room.state
        });

        subject.next({
          type: "CHANGE_ROUTE",
          routeName: "lobby"
        });
      });

      room.onStateChange.add(state => {
        console.log("Room State Change", state);
      });

      return subject;
    })
  );

export const getRoomListEpic = action$ =>
  action$.pipe(
    filter(action => action.type === "GET_ROOM_LIST"),

    mergeMap(action => {
      const subject = new Subject();
      // Call Colyseus client to get room list
      client.getAvailableRooms("standard", function(rooms, err) {
        if (err) {
          console.error(err);
        } else {
          console.log("Room from server", rooms);
          console.log(action, action$, subject);
          rooms.push({id: "1", name: "Test1"})
          rooms.push({id: "2", name: "Test2"})
          rooms.push({id: "3", name: "Test3"})

          subject.next({
            type: "UPDATE_ROOM_LIST",
            rooms
          });
        }
      })
      return subject;
    })
  );
