// reducer/GameStateReducer.js
import {
  GAME_STATE_NEW_ROOM,
  GAME_STATE_PLAYER_JOIN,
  GAME_CONTROL_X_UPDATE,
  GAME_CONTROL_Y_UPDATE
} from "action/GameActions";
import update from "react-addons-update";

const gamestate = function(
  state = {
    leftControl: {
      x: 0,
      y: 0
    },
    rightControl: {
      x: 0,
      y: 0
    }
  },
  action
) {
  // console.log(action.type, action);
  switch (action.type) {
    case GAME_STATE_PLAYER_JOIN:
      return update(state, {
        step: {
          $set: action.payload.step
        },
        players: {
          $set: action.payload.players
        },
        roominfo: {
          $set: action.payload.roominfo
        }
      });
    case GAME_CONTROL_X_UPDATE:
      if (action.gameControl === "leftControl") {
        return update(state, {
          leftControl: {
            x: { $set: action.newX }
          }
        });
      } else {
        return update(state, {
          rightControl: {
            x: { $set: action.newX }
          }
        });
      }
    case GAME_CONTROL_Y_UPDATE:
      if (action.gameControl === "leftControl") {
        return update(state, {
          leftControl: { y: { $set: action.newY } }
        });
      } else {
        return update(state, {
          rightControl: { y: { $set: action.newY } }
        });
      }

    default:
      return state;
  }
};

export default gamestate;
